mlflow/models/evaluation/base.py

```diff
@@ -151,6 +156,194 @@ def __str__(self):
         return "EvaluationMetric(" + ", ".join(parts) + ")"
 
 
+# NB: we need this function because we cannot modify the signature of
+# a class's __call__ method after the class has been defined.
+# This is also useful to distinguish between the metric signatures with different eval_fn signatures
+def dynamically_generate_eval_metric(eval_fn, strict_signature=False):
+    """
+    Dynamically generate a GenAIEvaluationMetric class that can be used to evaluate the metric
+    on the given input data. The generated class is callable with a __call__ method that
+    takes the arguments specified in the signature of the eval_fn function.
+
+    Args:
+        eval_fn: the evaluation function of the EvaluationMetric.
+        strict_signature: (Optional) Whether the metric follows a strict signature, if True then
+            the eval_fn must follow below signature:
+
+                .. code-block:: python
+
+                    def eval_fn(
+                        predictions: "pd.Series",
+                        metrics: Dict[str, MetricValue],
+                        inputs: "pd.Series",
+                        *args,
+                    ) -> MetricValue:
+                        pass
+
+            When generating a metric from `make_genai_metric`, this should be set to True.
+            Default to False.
+
+    Returns:
+        A dynamically generated callable CallableEvaluationMetric class.
+    """
+    from mlflow.metrics.base import MetricValue
+
+    if strict_signature:
+        allowed_kwargs_names = [
+            param_name
+            for param_name in inspect.signature(eval_fn).parameters.keys()
+            if param_name not in ["predictions", "metrics", "inputs"]
+        ]
+
+        def genai_call_method(
+            self,
+            *,
+            predictions: Union[pd.Series, str, list],
+            inputs: Union[pd.Series, str, list],
+            metrics: Optional[Dict[str, MetricValue]] = None,
```

Comment:
What is the use cases for passing this `metrics` argument?

Reply:
Ahh seems like it's never used :) I can remove it, but it exists in the underlying eval_fn signature, so wondering should we keep the consistency in case we make any updates in the future?

Reply:
I see, why do we need to keep consistency with `eval_fn`? Isn't it only used internally?

Reply:
That's true :) Let me remove it

Reply:
Emm is it possible that metrics is used inside `eval_fn` afterwards (now we can just pass None), and then this`__call__` method is broken?